0005-R-pubsub-delayed-scheduling.md

```diff
@@ -0,0 +1,156 @@
+
+* What is deliberately *not* in scope?
+
+Using component specific implementations is deliberately not in scope, as it offers no tangible benefits and carries heavy maintenance burden for contributors, approvers and maintainers.
+
+* What alternatives have been considered, and why do they not solve the problem?
+
+The only alternative for users is to create separate topics for delayed messages together with boilerplate code to orchestrate sending/receiving messages from said topics within their code, and then chaining the message flow to the correct topics after the conditions for sending delayed messages have been met. In addition, user code will need to become aware of Dapr specific mechanisms and mirror it in user code.
+
+* What advantages / disadvantages does this proposal have?
+
+This proposal allows delayed message scheduling to be enabled for:
+
+1. Every pub/sub component supported by Dapr without requiring component specific changes
+2. Messages sent using Dapr CloudEvents
+3. Messages sent using user provided CloudEvents
+4. Messages sent using raw payloads
+5. Message sent/received by HTTP and gRPC subscribers
+
+## Implementation Details
+
+### Design
+
+The Dapr runtime will use an intermediary topic to store messages that are scheduled for late delivery. Users will specify their intent for a message to be delayed using the existing metadata mechanism (specified as a query param on HTTP1.1 or on the request object for gRPC), the same way we use metadata for TTLs in pub/sub and for many other features. The proposed metadata key to be used is `scheduledSendTime` with the value being a date time string adhering to the ISO-8601 format, the same format supported by Dapr actors reminders and timers.
+
+HTTP example:
+
+```bash
+curl -X "POST" http://localhost:3500/v1.0/publish/pubsub/singularity?metadata.scheduledSendTime=2045-11-05T08:15:30-05:00 -H "Content-Type: application/json" -d '{"execute-order": "66"}'
+```
+
+Upon receiving a delayed message, the Dapr runtime will examine the due date for the message and publish it to the target topic if time is due. If the time isn't due, the Dapr runtime will hold the message until the time is right to send it without ACKing back to the broker. If the Dapr instance crashes, the message will remain in the broker and be consumed by a different instance.
+
+#### Configuring delayed messaging
+
+Users can enable delayed messaging by adding the following pub/sub component metadata (other fields specific for Redis removed for brevity):
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: pubsub
+spec:
+  type: pubsub.redis
+  version: v1
+  metadata:
+  - name: "enableDelayedMessaging"
+    value: true
```

Comment:
When this flag is set to true, we will be using the same `pubsub` with a different topic i.e. `<app>-delayed` to store the delayed messages and then later process it and send it to the correct topic in the `pubsub`?

Reply:
Yes, only if a publish reauest containing the metadata key for a scheduled event is present. Otherwise it's delivered to the target topic as is.